aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1387 @@
+    /// - `Err(PanicError)` if there was an invalid state (e.g., inconsistent status and flag)
+    ///
+    /// # Note
+    /// Each successful add_stall must be balanced by a corresponding remove_stall call.
+    pub(crate) fn add_stall(&self) -> Result<bool, PanicError> {
+        if self.num_stalls.fetch_add(1, Ordering::SeqCst) == 0 {
+            // Acquire write lock for (non-monitor) shortcut modifications.
+            let inner_status = self.inner_status.lock();
+
+            // num_stalls updates are not under the lock, so need to re-check (otherwise
+            // a different remove_stall might have already decremented the count).
+            if self.num_stalls.load(Ordering::Relaxed) == 0 {
+                return Ok(false);
+            }
+
+            let dependency_flag = self.dependency_shortcut.load(Ordering::Relaxed);
+            if !dependency_flags::is_valid(dependency_flag) {
+                return Err(code_invariant_error(format!(
+                    "Invalid dependency flag {dependency_flag} in add_stall",
+                )));
+            }
+
+            match (inner_status.pending_scheduling(), dependency_flag) {
+                (Some(0), dependency_flags::DEFER) => {
+                    // Adding a stall requires being recorded in aborted depedencies in scheduler_v2,
+                    // which in turn only happens in the scheduler after a successful abort (that must
+                    // increment the incarnation of the status).
+                    return Err(code_invariant_error("0-th incarnation in add_stall"));
+                },
+                (Some(_), dependency_flags::DEFER) => {
+                    self.execution_queue_manager
+                        .remove_from_schedule(self.txn_idx);
+                    // Shortcut not affected.
+                },
+                (None, dependency_flags::SAFE) => {
+                    // May not update SAFE flag at a future incorrect time (i.e. ABA), as observing
+                    // num_stalls = 0 under status is required to set SAFE flag, but impossible
+                    // until the corresponding remove_stall (that starts only after add_stall finishes).
+                    self.dependency_shortcut
+                        .store(dependency_flags::DEFER, Ordering::Relaxed);
+                },
+                (None, dependency_flags::EXECUTING | dependency_flags::DEFER) => {
+                    // Executing or aborted: shortcut not affected.
+                },
+                (Some(_), dependency_flags::SAFE | dependency_flags::EXECUTING) => {
+                    return Err(code_invariant_error(
+                        "Inconsistent status and dependency shortcut in add_stall",
+                    ));
+                },
+                (_, unsupported_flag_value) => {
+                    unreachable!("Unsupported flag value {unsupported_flag_value} in add_stall");
```

Comment:
Group the cases by the inner_status.pending_scheduling() return value.

Reply:
done